File: src/main/java/org/mitallast/queue/crdt/registry/DefaultCrdtRegistry.java
package org.mitallast.queue.crdt.registry;

import com.google.inject.Inject;
import com.google.inject.assistedinject.Assisted;
import gnu.trove.map.TLongObjectMap;
import gnu.trove.map.hash.TLongObjectHashMap;
import javaslang.control.Option;
import org.mitallast.queue.crdt.Crdt;
import org.mitallast.queue.crdt.commutative.GCounter;
import org.mitallast.queue.crdt.commutative.LWWRegister;
import org.mitallast.queue.crdt.replication.Replicator;

import java.util.concurrent.locks.ReentrantLock;

public class DefaultCrdtRegistry implements CrdtRegistry {

    private final int index;
    private final long replica;
    private final Replicator replicator;
    private final ReentrantLock lock = new ReentrantLock();

    private TLongObjectMap<Crdt> crdtMap = new TLongObjectHashMap<>();

    @Inject
    public DefaultCrdtRegistry(
        @Assisted int index,
        @Assisted long replica,
        @Assisted Replicator replicator
    ) {
        this.index = index;
        this.replica = replica;
        this.replicator = replicator;
    }

    @Override
    public int index() {
        return index;
    }

    @Override
    public boolean createLWWRegister(long id) {
        lock.lock();
        try {
            if (crdtMap.containsKey(id)) {
                return false;
            }
            crdtMap.put(id, new LWWRegister(id, replicator));
            return true;
        } finally {
            lock.unlock();
        }
    }

    @Override
    public boolean createGCounter(long id) {
        lock.lock();
        try {
            if (crdtMap.containsKey(id)) {
                return false;
            }
            crdtMap.put(id, new GCounter(id, replica, replicator));
            return true;
        } finally {
            lock.unlock();
        }
    }

    @Override
    public boolean remove(long id) {
        lock.lock();
        try {
            if (crdtMap.containsKey(id)) {
                return false;
            }
            crdtMap.remove(id);
            return true;
        } finally {
            lock.unlock();
        }
    }

    @Override
    public Crdt crdt(long id) {
        Crdt crdt = crdtMap.get(id);
        if (crdt == null) {
            throw new IllegalArgumentException("CRDT " + id + " not registered");
        }
        return crdt;
    }

    @Override
    public Option<Crdt> crdtOpt(long id) {
        return Option.of(crdtMap.get(id));
    }

    @Override
    @SuppressWarnings("unchecked")
    public <T extends Crdt> T crdt(long id, Class<T> type) {
        Crdt crdt = crdt(id);
        if (type.isInstance(crdt)) {
            return (T) crdt;
        } else {
            throw new IllegalArgumentException("CRDT " + id + " does not " + type.getSimpleName() + ", actual " + crdt.getClass().getSimpleName());
        }
    }

    @Override
    @SuppressWarnings("unchecked")
    public <T extends Crdt> Option<T> crdtOpt(long id, Class<T> type) {
        Crdt crdt = crdtMap.get(id);
        if (crdt == null) {
            return Option.none();
        }
        if (type.isInstance(crdt)) {
            return Option.some((T) crdt);
        } else {
            throw new IllegalArgumentException("CRDT " + id + " does not " + type.getSimpleName() + ", actual " + crdt.getClass().getSimpleName());
        }
    }
}

File: src/main/java/org/mitallast/queue/crdt/registry/DefaultCrdtRegistry.kt
package org.mitallast.queue.crdt.registry

import com.google.inject.Inject
import com.google.inject.assistedinject.Assisted
import gnu.trove.map.hash.TLongObjectHashMap
import javaslang.control.Option
import org.mitallast.queue.crdt.Crdt
import org.mitallast.queue.crdt.commutative.GCounter
import org.mitallast.queue.crdt.commutative.GSet
import org.mitallast.queue.crdt.commutative.LWWRegister
import org.mitallast.queue.crdt.commutative.OrderedGSet
import org.mitallast.queue.crdt.replication.Replicator
import java.util.concurrent.locks.ReentrantLock

@Suppress("UNCHECKED_CAST")
class DefaultCrdtRegistry @Inject constructor(
    @param:Assisted private val index: Int,
    @param:Assisted private val replica: Long,
    @param:Assisted private val replicator: Replicator
) : CrdtRegistry {
    private val lock = ReentrantLock()
    private val crdtMap = TLongObjectHashMap<Crdt>()

    override fun index(): Int = index

    override fun createLWWRegister(id: Long): Boolean {
        lock.lock()
        try {
            if (crdtMap.containsKey(id)) {
                return false
            }
            crdtMap.put(id, LWWRegister(id, replicator))
            return true
        } finally {
            lock.unlock()
        }
    }

    override fun createGCounter(id: Long): Boolean {
        lock.lock()
        try {
            if (crdtMap.containsKey(id)) {
                return false
            }
            crdtMap.put(id, GCounter(id, replica, replicator))
            return true
        } finally {
            lock.unlock()
        }
    }

    override fun createGSet(id: Long): Boolean {
        lock.lock()
        try {
            if (crdtMap.containsKey(id)) {
                return false
            }
            crdtMap.put(id, GSet(id, replicator))
            return true
        } finally {
            lock.unlock()
        }
    }

    override fun createOrderedGSet(id: Long): Boolean {
        lock.lock()
        try {
            if (crdtMap.containsKey(id)) {
                return false
            }
            crdtMap.put(id, OrderedGSet(id, replica, replicator))
            return true
        } finally {
            lock.unlock()
        }
    }

    override fun remove(id: Long): Boolean {
        lock.lock()
        try {
            if (crdtMap.containsKey(id)) {
                return false
            }
            crdtMap.remove(id)
            return true
        } finally {
            lock.unlock()
        }
    }

    override fun crdt(id: Long): Crdt {
        return crdtMap.get(id) ?: throw IllegalArgumentException("CRDT $id not registered")
    }

    override fun crdtOpt(id: Long): Option<Crdt> {
        return Option.of(crdtMap.get(id))
    }

    override fun <T : Crdt> crdt(id: Long, type: Class<T>): T {
        val crdt = crdt(id)
        return if (type.isInstance(crdt)) {
            crdt as T
        } else {
            throw IllegalArgumentException("CRDT " + id + " does not " + type.simpleName + ", actual " + crdt.javaClass.simpleName)
        }
    }

    override fun <T : Crdt> crdtOpt(id: Long, type: Class<T>): Option<T> {
        val crdt = crdtMap.get(id) ?: return Option.none()
        return if (type.isInstance(crdt)) {
            Option.some(crdt as T)
        } else {
            throw IllegalArgumentException("CRDT " + id + " does not " + type.simpleName + ", actual " + crdt.javaClass.simpleName)
        }
    }
}
